### PR TITLE
Adding Zcb implementation

### PR DESCRIPTION
**Pull request by junambi**
Zcb implementation for review and merger.

### CODE REVIEW

**Comment by Timmmm** [comment]
Following the discussion in Zoom I had a closer look at this and almost all of these instructions map very clearly to existing instructions. Only 2 don't map cleanly and would maybe be clearer reimplemented.

* `c.lbu` -> `lbu`
* `c.lhu` -> `lhu`
* `c.lh` -> `lh`
* `c.sb` -> `sb`
* `c.sh` -> `sh`
* `c.zext.b` -> reimplement?
* `c.sext.b` -> `sext.b`
* `c.zext.h` -> `zext.h`
* `c.sext.h` -> `sext.h`
* `c.not` -> reimplement?
* `c.mul` -> `mul`

For example instead of:

```
function clause execute (C_MUL(rsd,rs2)) = {
  let rd = creg2reg_idx(rsd);
  let rs = creg2reg_idx(rs2);

  let result_wide = to_bits(2 * sizeof(xlen), signed(X(rd)) * signed(X(rs)));
  X(rd) = result_wide[(sizeof(xlen) - 1) .. 0];
  RETIRE_SUCCESS
}
```

Use:

```
function clause execute (C_MUL(rsd,rs2)) = {
  let rd = creg2reg_idx(rsd);
  let rs = creg2reg_idx(rs2);

  execute(MUL((rs, rd, rd, false, true, true))
}
```

Primitive obsession notwithstanding, I think this is easier to understand, matches the existing C instructions and is less error prone. The intention of this extension is just to have shorter decodings, so the implementation *should* match.

For example in this case (which I picked randomly!), reimplementing it introduced a bug, whereas just mapping it would not have. I think that's probably a strong enough argument on its own, given the lack of testing.

**Comment by jrtc27** [comment]
> Following the discussion in Zoom I had a closer look at this and almost all of these instructions map very clearly to existing instructions. Only 2 don't map cleanly and would maybe be clearer reimplemented.
> 
> * `c.lbu` -> `lbu`
> * `c.lhu` -> `lhu`
> * `c.lh` -> `lh`
> * `c.sb` -> `sb`
> * `c.sh` -> `sh`
> * `c.zext.b` -> reimplement?
> * `c.sext.b` -> `sext.b`
> * `c.zext.h` -> `zext.h`
> * `c.sext.h` -> `sext.h`
> * `c.not` -> reimplement?
> * `c.mul` -> `mul`
> 
> For example instead of:
> 
> ```
> function clause execute (C_MUL(rsd,rs2)) = {
>   let rd = creg2reg_idx(rsd);
>   let rs = creg2reg_idx(rs2);
> 
>   let result_wide = to_bits(2 * sizeof(xlen), signed(X(rd)) * signed(X(rs)));
>   X(rd) = result_wide[(sizeof(xlen) - 1) .. 0];
>   RETIRE_SUCCESS
> }
> ```
> 
> Use:
> 
> ```
> function clause execute (C_MUL(rsd,rs2)) = {
>   let rd = creg2reg_idx(rsd);
>   let rs = creg2reg_idx(rs2);
> 
>   execute(MUL((rs, rd, rd, false, true, true))
> }
> ```
> 
> Primitive obsession notwithstanding, I think this is easier to understand, matches the existing C instructions and is less error prone. The intention of this extension is just to have shorter decodings, so the implementation _should_ match.
> 
> For example in this case (which I picked randomly!), reimplementing it introduced a bug, whereas just mapping it would not have. I think that's probably a strong enough argument on its own, given the lack of testing.

Yes. See https://github.com/riscv/sail-riscv/pull/255#discussion_r1169123796.

**Comment by martinberger** [comment]
I just asked Tariq Kurd, the original author of the spec what he thinks we should do:

1. Simply forward to the corresponding non-compressed instruction.
2. Implement the spec faithfully to RISC-V Instruction Set Manual (so not forwarding).

Tariq said we should do (1), that was always the intent - and demonstrates that the instructions are trivial. This is also how the other compressed instructions are implemented, e.g. `c.addi`

**Comment by abukharmeh** [comment]
Same thoughts https://github.com/riscv/sail-riscv/pull/255#discussion_r1170505749

**Comment by junambi** [comment]
c.lbu -> lbu                             :: (Is implemented as part of LOAD), i do not see a separate LBU implementation
    c.lhu -> lhu                             :: (Is implemented as part of LOAD), i do not see a separate LHU implementation
    c.lh -> lh                                 :: (Is implemented as part of LOAD), i do not see a separate LH implementation
    c.sb -> sb                               :: (Is implemented as part of STORE), i do not see a separate SB implementation
    c.sh -> sh                               :: (Is implemented as part of STORE), i do not see a separate SH implementation
    c.zext.b -> reimplement?        :: Is the current implementation fine for this
    c.sext.b -> sext.b                    :: Will redo as suggested reusing SEXT_B
    c.zext.h -> zext.h                    :: Will redo as suggested reusing ZEXT_H
    c.sext.h -> sext.h                    :: Will redo as suggested reusing SEXT_H
    c.not -> reimplement?        :: Is the current implementation fine for this
    c.mul -> mul                           :: Will redo as suggested reusing MUL

Just for my understanding, can one of you please clarify the following,
1. What are the advantages of using the existing instruction implementation, will it not affect the performance of the sail code
2. In HW, having to implement other instructions (except the base) to support compressed instructions, will it not affect the area, power and performance of the system as a whole / is it required for HW to implement many other extensions to support only the base and compressed instructions.

**Comment by junambi** [comment]
Thank you Martin,
Are the other listed implementation in my previous comment, good to implement.

Thank you.

**Comment by Timmmm** [comment]
> In HW, having to implement other instructions (except the base) to support compressed instructions, will it not affect the area, power and performance of the system as a whole / is it required for HW to implement many other extensions to support only the base and compressed instructions.

These instructions are only supported if the instructions they map to are are implemented. That was the bug in your `c.mul` implementation that I mentioned: `c.mul` is not supported if `mul` is not implemented.

**Comment by junambi** [comment]
Yes Tim, i understood that part. 

I was trying to understand, does this not put a limitation on the implementation, as someone could choose to implement only the compressed instructions. 

Is this even allowed/supported.

Thank you.

**Comment by Timmmm** [comment]
> as someone could choose to implement only the compressed instructions.

They can't. The ones that are mapped to optional instructions require those instructions to be implemented. For example here's what the spec says about `c.mul`:

image

Your `c.mul` implementation was missing the check for this, but the `mul` implementation has it:

```
function clause execute (MUL(rs2, rs1, rd, high, signed1, signed2)) = {
  if haveMulDiv() | haveZmmul() then {
```

**Comment by jrtc27** [comment]
... and there's no sensible use case for it, either; if you have the hardware to do a c.mul, you might as well use it to do a mul too. Decode is very little cost compared to the actual multiplier.

**Comment by junambi** [comment]
Modified the implementation and pushed the file, please review and comment.

File: model/riscv_insts_zcb.sail [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+mapping clause assembly = C_LBU(uimm, rsc, rdc)
+  <-> "c.lbu" ^ spc() ^ creg_name(rdc) ^ sep() ^ creg_name(rsc) ^ sep() ^ hex_bits_5(uimm @ 0b000)
+/* ****************************************************************** */
+union clause ast = C_LHU : (bits(1), cregidx, cregidx)
+
+mapping clause encdec_compressed = C_LHU(ui5, rs1, rd)
+  <-> 0b100 @ 0b001 @ rs1 : cregidx @ 0b0 @ ui5 : bits(1) @ rd : cregidx @ 0b00
+
+function clause execute (C_LHU(uimm, rsc, rdc)) = {
+  let immext : xlenbits = zero_extend(uimm);
```
Comment (Timmmm) [inline model/riscv_insts_zcb.sail:102]: Also unused. Please check the other instructions too.

File: model/riscv_insts_zcb.sail [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+  let imm : bits(12) = zero_extend(uimm);
```
Comment (Timmmm) [inline model/riscv_insts_zcb.sail:103]: Also wrong - should be `zero_extend(ui1 @ 0b0);`

Please check the other instructions too.

File: model/riscv_insts_zcb.sail [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+/* Sail Model for the c.mul encoding group                            */
+/* ------------------------------------------------------------------ */
+
+/* the assembly abstract syntax tree (AST) clause for the CMUL instructions */
+union clause ast = C_MUL : (cregidx,cregidx)
+
+/* the encode/decode mapping between AST elements and 16-bit half words */
+
+mapping clause encdec_compressed = C_MUL(rsd, rs2)
+  <-> 0b100 @ 0b111 @ rsd : cregidx @ 0b00 @ rs2 : cregidx @ 0b01
```
Comment (Timmmm) [inline model/riscv_insts_zcb.sail:260]: `@ 0b00 @` should be `@ 0b10 @`

**Comment by jrtc27** [comment]
Given these review comments, are you testing your code before pushing it for review, and if so how? Because basic things like decode patterns should be caught by even the most basic testing.

Also you've still deleted a bunch of V-related changes from Makefile which causes the build to fail in CI. So I don't think you could possibly have tested this code given it won't build.

**Comment by junambi** [comment]
Thank you Tim and Jessica for the comments, i have updated the Makefile and the Zcb model.

File: Makefile [hunk 4]
```diff
@@ -30,6 +30,9 @@ SAIL_DEFAULT_INST += riscv_insts_zbc.sail
 SAIL_DEFAULT_INST += riscv_insts_zbs.sail
 
 SAIL_DEFAULT_INST += riscv_insts_zfh.sail
+
+SAIL_DEFAULT_INST += riscv_insts_zcb.sail
```
Comment (arichardson) [inline Makefile:34]: It would be good to keep this list sorted alphabetically, i.e. move it up one.

File: model/riscv_insts_zcb.sail [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+  let rs = creg2reg_idx(rsc);
+  execute(LOAD(imm, rs, rd, true, BYTE, false, false))
+}
+
+mapping clause assembly = C_LBU(uimm, rsc, rdc)
+  <-> "c.lbu" ^ spc() ^ creg_name(rdc) ^ sep() ^ creg_name(rsc) ^ sep() ^ hex_bits_5(uimm @ 0b000)
+/* ****************************************************************** */
+union clause ast = C_LHU : (bits(1), cregidx, cregidx)
+
+mapping clause encdec_compressed = C_LHU(ui1, rs1, rd)
```
Comment (arichardson) [inline model/riscv_insts_zcb.sail:97]:
```suggestion
mapping clause encdec_compressed = C_LHU(ui1 @Ob0, rs1, rd)
```

Immediate is not scaled in current code. I assume this means there are no tests?

**Comment by Timmmm** [comment]
> Thank you Tim and Jessica for the comments, i have updated the Makefile and the Zcb model.

Looks like you didn't fix the issues with the immediates though?

**Comment by jjscheel** [comment]
> Given these review comments, are you testing your code before pushing it for review, and if so how? Because basic things like decode patterns should be caught by even the most basic testing.

Hi, @jrtc27, fair question.  I asked @junambi to post his current PR work early because @martinberger had interest in the work and we wanted to make sure that we didn't duplicating previous effort.

File: Makefile (hunk 4 above)
Comment (junambi) [reply Makefile]: This has been updated.

File: model/riscv_insts_zcb.sail [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+
+/* ****************************************************************** */
+union clause ast = C_LBU : (bits(2), cregidx, cregidx)
+
+mapping clause encdec_compressed = C_LBU(ui65, rs1, rd)
+  <-> 0b100 @ 0b000 @ rs1 : cregidx @ ui65 : bits(2) @ rd : cregidx @ 0b00
+
+function clause execute (C_LBU(uimm, rsc, rdc)) = {
+  let immext : xlenbits = zero_extend(uimm);
+  let imm : bits(12) = zero_extend(uimm);
```
[the remark this answers is not in the document]
Comment (jrtc27) [reply model/riscv_insts_zcb.sail:85]: Swizzle the bits in decode not execute

**Comment by jjscheel** [comment]
@martinberger, what is the status of your work?  I can't tell if a PR was submitted.

**Comment by Timmmm** [comment]
There isn't a PR. I thought Martin added our implementation to a comment but I can't see it now.  I'll make a new PR.

**Comment by jjscheel** [comment]
Thanks, @Timmmm and @martinberger!

**Pull request closed without merging**